STM32WB0: add DMA support

mathieuchopstm opened the pull request:
Add support for DMA+DMAMUX on STM32WB0 series hardware and updates the boards' YAML accordingly; also add a test overlay for Nucleo-WB09KE in relevant DMA tests so that correct behavior can be verified.

mathieuchopstm committed:
dts: bindings: dma: make DMAMUX IRQ optional

Remove the "required: true" attribute from the STM32 DMAMUX binding.
This is required for STM32WB0 series where the DMAMUX has no interrupt line

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
drivers: dma: stm32: add support for STM32WB0

Add support for STM32WB0 series in the relevant drivers and Kconfig.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
dts: arm: st: wb0: add DMA and DMAMUX nodes

Add device tree nodes corresponding to DMA and DMAMUX peripherals to
STM32WB0 series DTSI.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
tests: dma: chan_blen_transfer: add overlay for Nucleo-WB09KE

Add an overlay to the test so that it can run on STM32WB0 series hardware.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
tests: dma: loop_transfer: add overlay for Nucleo-WB09KE

Add an overlay to the test so that it can run on STM32WB0 series hardware.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm committed:
boards: st: wb0: add DMA to supported feature set

Add DMA to supported feature set on existing STM32WB0 boards.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

erwango commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -76,6 +76,14 @@ uint32_t table_ll_channel[] = {
+ * STM32WB0 series DMAMUX lacks synchronization inputs and request generator.
+ * For this reason, the following functions are not present in STM32CubeWB0 LL.
+ *
+ * @note As of writing, these functions are not used by the STM32 driver.
```
If not used, then let's clean them up.

mathieuchopstm replied:
Done.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -780,6 +780,7 @@ static void dma_stm32_config_irq_0(const struct device *dev)
 	ARG_UNUSED(dev);
 
 	DMA_STM32_IRQ_CONNECT(0, 0);
+#if !defined(CONFIG_SOC_SERIES_STM32WB0X) /* Only 1 IRQ on STM32WB0 */
```
Can't use `#if DT_INST_IRQ_HAS_IDX(0, 1)` ?

mathieuchopstm replied:
Indeed, it should work. Changing.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -30,7 +30,7 @@ LOG_MODULE_REGISTER(dma_stm32, CONFIG_DMA_LOG_LEVEL);
 
 #if DT_NODE_HAS_STATUS_OKAY(DT_DRV_INST(0))
-#if DT_INST_IRQ_HAS_IDX(0, 7)
+#if DT_INST_IRQ_HAS_IDX(0, 7) || defined(CONFIG_SOC_SERIES_STM32WB0X)
```
Some information (could be in commit message) about why a `CONFIG_SOC_SERIES_STM32WB0X` specific rule is needed would be welcome.

mathieuchopstm replied:
Added a comment above.

erwango commented on drivers/dma/dmamux_stm32.c:
```diff
@@ -56,41 +56,12 @@ struct dmamux_stm32_config {
-	LISTIFY(DT_INST_PROP(0, dma_channels), DMAMUX_CHANNEL, (,))
-};
-
 #if !defined(CONFIG_SOC_SERIES_STM32G0X) && \
 	!defined(CONFIG_SOC_SERIES_STM32C0X)
 #define dmamux_channel_typedef DMAMUX_Channel_TypeDef
 #else
 #define dmamux_channel_typedef const DMAMUX_Channel_TypeDef
 #endif
```
Probably this could be removed as well.

erwango reviewed (changes requested):
Otherwise LGTM

mathieuchopstm commented:
Patch V2:
* remove remaining unused piece of code in DMAMUX driver
* refactor DMA driver to reduce per-series crust and make intents more explicit
* use proper binding and `interrupts` property in DTS

mathieuchopstm committed:
drivers: dma: dmamux_stm32: remove unused SO and RGO functions

Functions related to synchronization inputs and request generators were
present in the DMAMUX driver, despite being completely unused.
Remove them for compatibility with WB0 which lacks this hardware.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>

mathieuchopstm commented:
Fixed wrong `#dma-cells` value in DTS + split DMAMUX driver modification to separate commit

gautierg-st commented on dts/arm/st/wb0/stm32wb0.dtsi:
```diff
@@ -217,6 +217,31 @@
+			compatible = "st,stm32-dma-v2bis";
+			#dma-cells = <2>;
+			reg = <0x48700000 256>;
+			clocks = <&rcc STM32_CLOCK_BUS_AHB0 (1 << 0)>;
```
Now that #79683 has been merged, this node should be updated

mathieuchopstm replied:
Done + added a commit to update nodes missed by the PR (I2C/SPI)

gautierg-st commented on dts/arm/st/wb0/stm32wb0.dtsi:
```diff
@@ -217,6 +217,31 @@
+			/* `clocks` property is identical between DMA and DMAMUX
+			 * because they share a single common bit in RCC registers
+			 */
+			clocks = <&rcc STM32_CLOCK_BUS_AHB0 (1 << 0)>;
```
Now that #79683 has been merged, this node should be updated

mathieuchopstm replied:
Done

mathieuchopstm committed:
dts: arm: st: wb0: use STM32_CLOCK everywhere

PR #79683 missed a few nodes introduced while it was under review.
Replace the remaining raw values with STM32_CLOCK in WB0 DTSI files.

Signed-off-by: Mathieu Choplain <mathieu.choplain@st.com>
